pkg/policy/l4.go

```diff
@@ -1084,13 +1111,280 @@ func addL4Filter(policyCtx PolicyContext,
 		}
 	}
 
-	resMap[key] = existingFilter
+	resMap.Upsert(p.Port, uint16(p.EndPort), string(proto), existingFilter)
 	return nil
 }
 
-// L4PolicyMap is a list of L4 filters indexable by protocol/port
-// key format: "port/proto"
-type L4PolicyMap map[string]*L4Filter
+// L4PolicyMap is a list of L4 filters indexable by port/endport/protocol
+type L4PolicyMap interface {
+	Upsert(port string, endPort uint16, protocol string, l4 *L4Filter)
+	Delete(port string, endPort uint16, protocol string)
+	ExactLookup(port string, endPort uint16, protocol string) *L4Filter
+	LongestPrefixMatch(port string, protocol string) *L4Filter
+	Detach(selectorCache *SelectorCache)
+	IngressCoversContext(ctx *SearchContext) api.Decision
+	EgressCoversContext(ctx *SearchContext) api.Decision
+	ForEach(func(l4 *L4Filter) bool)
+	Equals(t *testing.T, bMap L4PolicyMap) bool
+	Diff(t *testing.T, expectedMap L4PolicyMap) string
```

Comment:
Why does production code link against `*testing.T`? Is it possible to move that into a testing wrapper?

Reply:
> Why does production code link against *testing.T? Is it possible to move that into a testing wrapper?

Because those methods are used by other packages.

Reply:
> This PR does not introduce this dependency. It was introduced when L4PolicyMap became an interface.

I'm confused. This comment thread is responding to the change for `L4PolicyMap` to convert it to an interface and add the `testing.T` dependency, as well as the implementations for this new interface.

I understand it might be a little less convenient to move this testing-only functionality into a testing-only file, but that seems better than entangling the direct testing dependency into the prod code like this. There's probably a reasonable wrapper we could create, for instance if we moved `InitL4PolicyMap()` into the testing code and forced it to return an object that (a) embeds L4PolicyMap and (b) implements the additional testing-only functions.

Reply:
> I'm confused. This comment thread is responding to the change for L4PolicyMap to convert it to an interface and add the testing.T dependency, as well as the implementations for this new interface.

Ah sorry, my bad. It wasn't L4PolicyMap, but `MapState` that introduced a testing dependency into the package with its `Diff` method. 

> I guess this particular proposal doesn't fit with the suggestion for using builder pattern below, but there's probably a reasonable solution either by avoiding builder pattern or by creating a testing-only wrapper around the implementation with builder pattern.

Since the cat is already out of the bag I would prefer to remove the `testing` package dependency out in a different PR.

Reply:
Ah right, that makes more sense. Fair enough, we can follow up on this.